Make tool parameter options capable of referring to metadata file

qiagu opened the pull request:
## What did you do? 
- Extend tool parameter options to generate dynamic options from metadata file(s) of a referred dataset or dataset collection. 
- Refactor linter for dynamic options.


## Why did you make this change?
- The `from_dataset` is able to generate dynamic options from the main file(s) of a dataset or dataset_collection, while `data_meta` in `filter` can refer to some types of metadata via attribute `key`, but file type is not one of them, as stated in #8599.


## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

## For UI Components
- [ ] I've included a screenshot of the changes

qiagu committed:
Make tool parameter options capable of referring to metadata file

qiagu committed:
Make test tool for meta_file_key

qiagu committed:
Add simple doc for usage of meta_file_key

qiagu committed:
Fix tool styling

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -3519,6 +3533,11 @@ used to generate dynamic options.
     </xs:attribute>
+    <xs:attribute name="meta_file_key" type="xs:string">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">Documentation for meta_file_key</xs:documentation>
```
Can you replace this with actual docs?

qiagu replied:
Frankly, I'm not sure how to do this. I was a copycat. :)

bernt-matthias replied:
Just put text there describing how the attribute works.

Also thinking if it's worth to extend the linter https://github.com/galaxyproject/galaxy/blob/b52c78ac1e7ebf36e4ca735e6c6a53838260f15c/lib/galaxy/tool_util/linters/inputs.py#L31

bernt-matthias commented on test/functional/tools/options_from_metadata_file.xml:
```diff
@@ -0,0 +1,29 @@
+<tool id="options_from_metadata_file" name="Test for options from dataset metadata file" version="1.0.0" profile="21.01">
+  <command><![CDATA[
+#assert $input.metadata.species_chromosomes
```
Out of curiosity: What is the idea of this assertion?

qiagu replied:
Copied from another test tool. It doesn't seem to be harmful, but I will be happy to remove it.

bernt-matthias commented on test/functional/tools/options_from_metadata_file.xml:
```diff
@@ -0,0 +1,29 @@
+  ]]></command>
+  <inputs>
+    <param name="input" type="data" format="maf" label="MAF File"/>
+    <param name="species" type="select" optional="False" label="Select species for the input dataset" multiple="True">
```
```suggestion
    <param name="species" type="select" optional="false" label="Select species for the input dataset" multiple="true">
```

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
````diff
@@ -3430,6 +3430,20 @@ and demonstrates generating options from a dataset directly.
 </param>
 ```
 
+The ``meta_file_key`` can be used together with ``from_dataset``. In such cases,
````
Should we document that this works only with 21.01 / requires profile="21.01".

qiagu replied:
Sure. Will do.

qiagu committed:
Update test/functional/tools/options_from_metadata_file.xml

Co-authored-by: M Bernt <m.bernt@ufz.de>

qiagu committed:
Merge branch 'dev' into dynamic_options

qiagu committed:
Add test for options from metadata file of dataset_collection

qiagu committed:
Improve docs

bernt-matthias committed:
add more linting for select parameters

bernt-matthias committed:
docs for more attributes of <options>

bernt-matthias reviewed (changes requested):
Hi @qiagu I was so free to add a bit of linting to this PR .. would be great if someone could have a look. 

Missing is only the doc for the new attribute and the PR template.

bernt-matthias committed:
linter fixes

qiagu committed:
Add lint warning for meta_file_key and further improve documentation

qiagu committed:
Merge branch 'dev' into dynamic_options

qiagu committed:
Merge branch 'dynamic_options' of https://github.com/qiagu/galaxy into dynamic_options

bernt-matthias committed:
add tests for new linters

bernt-matthias committed:
select linters: allow option definition by filters

- options can be added by add_value (but this works apparently only
  if there are already values)
- data_meta is also documented to be able to generate values

bernt-matthias committed:
use a different file for test

to check if options are added

bernt-matthias committed:
fix linter

bernt-matthias committed:
fix import order

bernt-matthias committed:
debug: switch test files

bernt-matthias committed:
debug: switch test files cont.

bernt-matthias committed:
fix test

Despite lines in 4.maf like

```
e echTel1.scaffold_309554      3896 34801 +     42008 I
```

echTel1 seems not to be in the metadata table.

So I have chosen another one.

bernt-matthias commented on lib/galaxy/tools/parameters/dynamic_options.py:
```diff
@@ -669,7 +670,9 @@ def get_fields(self, trans, other_values):
             for dataset in datasets:
-                if not hasattr(dataset, 'file_name'):
+                if self.meta_file_key:
+                    dataset = getattr(dataset.metadata, self.meta_file_key, None)
```
I guess we should check this against `no_value` instead of `None`?

bernt-matthias reviewed (commented):
A general comment on the example: The data used for the select that is taken from the `species_chromosomes` file is also contained in the `species` metadata element: 

https://github.com/galaxyproject/galaxy/blob/cf5ee5aed0cc164b6471a525e0a3da2b71ae4c8a/lib/galaxy/datatypes/sequence.py#L301

So there may be easier ways to implement this specific example.

qiagu commented on lib/galaxy/tools/parameters/dynamic_options.py:
```diff
@@ -669,7 +670,9 @@ def get_fields(self, trans, other_values):
-                if not hasattr(dataset, 'file_name'):
+                if self.meta_file_key:
+                    dataset = getattr(dataset.metadata, self.meta_file_key, None)
+                if not dataset or not hasattr(dataset, 'file_name'):
```
I think this line can help check the `no_value`.

qiagu committed:
Merge branch 'dev' into dynamic_options

qiagu committed:
Make sure meta_file_key refers to MetadataFile

qiagu committed:
Make sure metadata file is not deleted

qiagu committed:
Remove extra sign

qiagu committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into dynamic_options

bernt-matthias reviewed (approved):
Since I co-authored a few commits it would be good if someone else could also review these changes.

In particular I'm unsure if the new lints should be errors / warning .. I have the feeling that the new ones are a bit more "aggressive" then usual (but maybe that is good).

nsoranzo commented on lib/galaxy/tool_util/linters/inputs.py:
```diff
@@ -29,20 +30,75 @@ def lint_inputs(tool_xml, lint_ctx):
+
+            # check if options are defined by exactly one possibility
+            if (dynamic_options is not None) + (len(options) > 0) + (len(select_options) > 0) != 1:
+                lint_ctx.error(f"Select parameter [{param_name}] options have to be defined by either 'option' tags, a 'option' tag or the attribute 'dynamic_options'.")
```
```suggestion
                lint_ctx.error(f"Select parameter [{param_name}] options have to be defined by either 'option' children elements, a 'options' element or the 'dynamic_options' attribute.")
```

qiagu committed:
Give lint error to ensure meta_file_key works with from_dataset exclusively

qiagu committed:
Update lib/galaxy/tool_util/linters/inputs.py

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

qiagu committed:
Fix test error

nsoranzo commented on lib/galaxy/tool_util/linters/inputs.py:
```diff
@@ -29,20 +30,72 @@ def lint_inputs(tool_xml, lint_ctx):
+                    lint_ctx.warn(f"Select parameter [{param_name}] options uses deprecated 'transform_lines' attribute.")
+
+            elif len(options) > 1:
+                lint_ctx.error(f"Select parameter [{param_name}] contains multiple options tags")
```
```suggestion
                lint_ctx.error(f"Select parameter [{param_name}] contains multiple options elements")
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -3486,37 +3509,42 @@ used to generate dynamic options.
     <xs:attribute name="from_dataset" type="xs:string">
       <xs:annotation>
-        <xs:documentation xml:lang="en">Documentation for from_dataset</xs:documentation>
+        <xs:documentation xml:lang="en">Determine options from the data set given in the referred input parameter. If 'meta_file_key' is given the options are determined from the data in the metadata file of the input.</xs:documentation>
```
```suggestion
        <xs:documentation xml:lang="en">Determine options from the dataset given in the referred input parameter. If 'meta_file_key' is given, the options are determined from the data in the metadata file of the input.</xs:documentation>
```

qiagu committed:
Update lib/galaxy/tool_util/linters/inputs.py

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

qiagu committed:
Update lib/galaxy/tool_util/xsd/galaxy.xsd

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

qiagu committed:
Fix linter test
